ibmcloud: e2e test for flexible instance profiles

sudharshanibm3 opened the pull request:
Added e2e  test cases for flexible instance profiles

Fixes: #1161

stevenhorsman commented on test/e2e/common.go:
```diff
@@ -62,6 +62,22 @@ func withPVCBinding(mountPath string, pvcName string) podOption {
 	}
 }
 
+func withContainerPort(port int) podOption {
+	return func(p *corev1.Pod) {
+		p.Spec.Containers[0].Ports = []corev1.ContainerPort{
+			{
+				ContainerPort: int32(port),
+			},
+		}
+	}
+}
+
```
It doesn't look like we are using `withContainerPort` here, so I don't think we need to add this function?

stevenhorsman commented on test/e2e/common_suite_test.go:
```diff
@@ -257,24 +266,47 @@ func (tc *testCase) run() {
 					}
 					t.Logf("Log output of peer pod:%s", LogString)
 				}
-
-				if tc.isAuth {
-					if err := getAuthenticatedImageStatus(ctx, client, tc.AuthImageStatus, *tc.pod); err != nil {
```
Did you mean to delete the authenticated image tests? If not, can you please restore this and rebase this branch

stevenhorsman commented:
Running locally I got a failure with one of the tests:
```
=== RUN   TestPodVMwithAnnotationsLargerCPU
=== RUN   TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test
=== RUN   TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test/Failed_to_Create_PodVM_with_Annotations_Larger_CPU
time="2023-09-08T04:00:57-07:00" level=info msg="Failed to Get Expected Error: Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : failed to get instance type based on vCPU and memory annotations: no instance type found for the given vcpus (3) and memory (12288): unknown"
    common_suite_test.go:284: Pod Failed to execute expected error message Failed to Create PodVM Instance
time="2023-09-08T04:00:57-07:00" level=info msg="Deleting pod annotations-too-big-cpu..."
--- FAIL: TestPodVMwithAnnotationsLargerCPU (5.56s)
    --- FAIL: TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test (5.56s)
        --- FAIL: TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test/Failed_to_Create_PodVM_with_Annotations_Larger_CPU (0.41s)
```

stevenhorsman commented on test/e2e/common_suite_test.go:
```diff
@@ -246,7 +255,14 @@ func (tc *testCase) run() {
 						}
 					}
+				}
+
+				if tc.isAuth {
+					if err := getAuthenticatedImageStatus(ctx, client, tc.AuthImageStatus, *tc.pod); err != nil {
+						t.Fatal(err)
+					}
 
+					t.Logf("PodVM has successfully reached %v state with authenticated Image - %v", tc.AuthImageStatus, os.Getenv("AUTHENTICATED_REGISTRY_IMAGE"))
```
Is there are reason that isAuth section was moved from it's original place below the `if tc.expectedPodLogString` check?

sudharshanibm3 replied:
No, this might happened while rebasing and resolving conflicts, and it got added back

stevenhorsman replied:
Can you move it to the original place please to keep the diff simpler. Thanks

sudharshanibm3 commented:
Hi @stevenhorsman ,
Rebased and Updated code with few changes now
```
Wait for the kata-remote runtimeclass be created
=== RUN   TestPodVMwithNoAnnotations
=== RUN   TestPodVMwithNoAnnotations/PodVMWithNoAnnotations_test
    common_suite_test.go:208: Expected Pod State: Running
    common_suite_test.go:209: Current Pod State: Running
=== RUN   TestPodVMwithNoAnnotations/PodVMWithNoAnnotations_test/PodVM_with_No_Annotation_is_created
time="2023-09-08T17:04:26+05:30" level=info msg="PodVM Created with bz2-2x8 Instance type successfully..."
time="2023-09-08T17:04:26+05:30" level=info msg="PodVM name: no-annotations"
time="2023-09-08T17:04:27+05:30" level=debug msg="Instance number: 0, Instance id: 0787_8ab48654-cddb-4e9b-a9b2-80e14965759c, Instance name: podvm-no-annotations-00e4811e"
time="2023-09-08T17:04:27+05:30" level=info msg="Deleting pod no-annotations..."
--- PASS: TestPodVMwithNoAnnotations (29.91s)
    --- PASS: TestPodVMwithNoAnnotations/PodVMWithNoAnnotations_test (29.91s)
        --- PASS: TestPodVMwithNoAnnotations/PodVMWithNoAnnotations_test/PodVM_with_No_Annotation_is_created (3.80s)
=== RUN   TestPodVMwithAnnotationsInstanceType
=== RUN   TestPodVMwithAnnotationsInstanceType/PodVMwithAnnotationsInstanceType_test
    common_suite_test.go:208: Expected Pod State: Running
    common_suite_test.go:209: Current Pod State: Running
=== RUN   TestPodVMwithAnnotationsInstanceType/PodVMwithAnnotationsInstanceType_test/PodVM_with_Annotation_is_created
time="2023-09-08T17:04:54+05:30" level=info msg="PodVM Created with cz2-2x4 Instance type successfully..."
time="2023-09-08T17:04:55+05:30" level=info msg="PodVM name: annotations-instance-type"
time="2023-09-08T17:04:55+05:30" level=debug msg="Instance number: 0, Instance id: 0787_7565e6b3-6908-4d7a-9b7a-0d750af24aa3, Instance name: podvm-annotations-instance-type-8a1e169b"
time="2023-09-08T17:04:56+05:30" level=info msg="Deleting pod annotations-instance-type..."
--- PASS: TestPodVMwithAnnotationsInstanceType (28.82s)
    --- PASS: TestPodVMwithAnnotationsInstanceType/PodVMwithAnnotationsInstanceType_test (28.82s)
        --- PASS: TestPodVMwithAnnotationsInstanceType/PodVMwithAnnotationsInstanceType_test/PodVM_with_Annotation_is_created (2.75s)
=== RUN   TestPodVMwithAnnotationsCPUMemory
=== RUN   TestPodVMwithAnnotationsCPUMemory/PodVMwithAnnotationsCPUMemory_test
    common_suite_test.go:208: Expected Pod State: Running
    common_suite_test.go:209: Current Pod State: Running
=== RUN   TestPodVMwithAnnotationsCPUMemory/PodVMwithAnnotationsCPUMemory_test/PodVM_with_Annotations_CPU_Memory_is_created
time="2023-09-08T17:05:23+05:30" level=info msg="PodVM Created with mz2-2x16 Instance type successfully..."
time="2023-09-08T17:05:23+05:30" level=info msg="PodVM name: annotations-cpu-mem"
time="2023-09-08T17:05:25+05:30" level=debug msg="Instance number: 0, Instance id: 0787_627e205e-8a3a-4183-8782-eac6c99d5370, Instance name: podvm-annotations-cpu-mem-d4a9748b"
time="2023-09-08T17:05:25+05:30" level=info msg="Deleting pod annotations-cpu-mem..."
--- PASS: TestPodVMwithAnnotationsCPUMemory (29.33s)
    --- PASS: TestPodVMwithAnnotationsCPUMemory/PodVMwithAnnotationsCPUMemory_test (29.33s)
        --- PASS: TestPodVMwithAnnotationsCPUMemory/PodVMwithAnnotationsCPUMemory_test/PodVM_with_Annotations_CPU_Memory_is_created (3.24s)
=== RUN   TestPodVMwithAnnotationsInvalidInstanceType
=== RUN   TestPodVMwithAnnotationsInvalidInstanceType/PodVMwithAnnotationsInvalidInstanceType_test
=== RUN   TestPodVMwithAnnotationsInvalidInstanceType/PodVMwithAnnotationsInvalidInstanceType_test/Failed_to_Create_PodVM_with_Annotations_Invalid_InstanceType
time="2023-09-08T17:05:32+05:30" level=info msg="Got Expected Error: Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : failed to verify instance type: requested instance type (\"bz2-2x4\") is not part of supported instance types list: unknown"
time="2023-09-08T17:05:32+05:30" level=info msg="Deleting pod annotations-invalid-instance-type..."
--- PASS: TestPodVMwithAnnotationsInvalidInstanceType (7.20s)
    --- PASS: TestPodVMwithAnnotationsInvalidInstanceType/PodVMwithAnnotationsInvalidInstanceType_test (7.20s)
        --- PASS: TestPodVMwithAnnotationsInvalidInstanceType/PodVMwithAnnotationsInvalidInstanceType_test/Failed_to_Create_PodVM_with_Annotations_Invalid_InstanceType (1.41s)
=== RUN   TestPodVMwithAnnotationsLargerMemory
=== RUN   TestPodVMwithAnnotationsLargerMemory/PodVMwithAnnotationsLargerMemory_test
=== RUN   TestPodVMwithAnnotationsLargerMemory/PodVMwithAnnotationsLargerMemory_test/Failed_to_Create_PodVM_with_Annotations_Larger_Memory
time="2023-09-08T17:05:39+05:30" level=info msg="Got Expected Error: Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : failed to get instance type based on vCPU and memory annotations: no instance type found for the given vcpus (2) and memory (18432): unknown"
time="2023-09-08T17:05:39+05:30" level=info msg="Deleting pod annotations-too-big-mem..."
--- PASS: TestPodVMwithAnnotationsLargerMemory (7.26s)
    --- PASS: TestPodVMwithAnnotationsLargerMemory/PodVMwithAnnotationsLargerMemory_test (7.26s)
        --- PASS: TestPodVMwithAnnotationsLargerMemory/PodVMwithAnnotationsLargerMemory_test/Failed_to_Create_PodVM_with_Annotations_Larger_Memory (1.43s)
=== RUN   TestPodVMwithAnnotationsLargerCPU
=== RUN   TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test
=== RUN   TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test/Failed_to_Create_PodVM_with_Annotations_Larger_CPU
time="2023-09-08T17:05:46+05:30" level=info msg="Got Expected Error: Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : failed to get instance type based on vCPU and memory annotations: no instance type found for the given vcpus (3) and memory (12288): unknown"
time="2023-09-08T17:05:47+05:30" level=info msg="Deleting pod annotations-too-big-cpu..."
--- PASS: TestPodVMwithAnnotationsLargerCPU (7.13s)
    --- PASS: TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test (7.13s)
        --- PASS: TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test/Failed_to_Create_PodVM_with_Annotations_Larger_CPU (1.31s)
PASS
ok      github.com/confidential-containers/cloud-api-adaptor/test/e2e   192.205s
```

stevenhorsman commented:
I ran this locally and all the tests passed except:
```
=== RUN   TestCreatePodWithSecret/SecretPeerPod_test/Secret_has_been_created_and_contains_data
    common_suite_test.go:324: cat: /etc/secret/username: No such file or directory

    common_suite_test.go:325: command terminated with exit code 1
time="2023-09-11T03:16:05-07:00" level=info msg="Deleting Secret... nginx-secret"
time="2023-09-11T03:16:05-07:00" level=info msg="Deleting pod nginx-secret-pod..."
--- FAIL: TestCreatePodWithSecret (40.63s)
    --- FAIL: TestCreatePodWithSecret/SecretPeerPod_test (40.63s)
        --- FAIL: TestCreatePodWithSecret/SecretPeerPod_test/Secret_has_been_created_and_contains_data (5.33s)
```
so all the flexible instance tests passed:
```
=== RUN   TestPodVMwithNoAnnotations
=== RUN   TestPodVMwithNoAnnotations/PodVMWithNoAnnotations_test
    common_suite_test.go:208: Expected Pod State: Running
    common_suite_test.go:209: Current Pod State: Running
=== RUN   TestPodVMwithNoAnnotations/PodVMWithNoAnnotations_test/PodVM_with_No_Annotation_is_created
time="2023-09-11T03:22:53-07:00" level=info msg="PodVM Created with bz2-2x8 Instance type successfully..."
time="2023-09-11T03:22:53-07:00" level=info msg="PodVM name: no-annotations"
time="2023-09-11T03:22:54-07:00" level=debug msg="Instance number: 0, Instance id: 0797_79cc7de4-f321-4c44-9701-b93548f7d91a, Instance name: podvm-no-annotations-6ee472fd"
time="2023-09-11T03:22:54-07:00" level=info msg="Deleting pod no-annotations..."
--- PASS: TestPodVMwithNoAnnotations (26.18s)
    --- PASS: TestPodVMwithNoAnnotations/PodVMWithNoAnnotations_test (26.18s)
        --- PASS: TestPodVMwithNoAnnotations/PodVMWithNoAnnotations_test/PodVM_with_No_Annotation_is_created (0.97s)
=== RUN   TestPodVMwithAnnotationsInstanceType
=== RUN   TestPodVMwithAnnotationsInstanceType/PodVMwithAnnotationsInstanceType_test
    common_suite_test.go:208: Expected Pod State: Running
    common_suite_test.go:209: Current Pod State: Running
=== RUN   TestPodVMwithAnnotationsInstanceType/PodVMwithAnnotationsInstanceType_test/PodVM_with_Annotation_is_created
time="2023-09-11T03:23:19-07:00" level=info msg="PodVM Created with cz2-2x4 Instance type successfully..."
time="2023-09-11T03:23:19-07:00" level=info msg="PodVM name: annotations-instance-type"
time="2023-09-11T03:23:20-07:00" level=debug msg="Instance number: 0, Instance id: 0797_b0e67e2e-797a-4328-bcff-f526fb97bb25, Instance name: podvm-annotations-instance-type-252b35ae"
time="2023-09-11T03:23:20-07:00" level=info msg="Deleting pod annotations-instance-type..."
--- PASS: TestPodVMwithAnnotationsInstanceType (26.14s)
    --- PASS: TestPodVMwithAnnotationsInstanceType/PodVMwithAnnotationsInstanceType_test (26.14s)
        --- PASS: TestPodVMwithAnnotationsInstanceType/PodVMwithAnnotationsInstanceType_test/PodVM_with_Annotation_is_created (0.95s)
=== RUN   TestPodVMwithAnnotationsCPUMemory
=== RUN   TestPodVMwithAnnotationsCPUMemory/PodVMwithAnnotationsCPUMemory_test
    common_suite_test.go:208: Expected Pod State: Running
    common_suite_test.go:209: Current Pod State: Running
=== RUN   TestPodVMwithAnnotationsCPUMemory/PodVMwithAnnotationsCPUMemory_test/PodVM_with_Annotations_CPU_Memory_is_created
time="2023-09-11T03:23:45-07:00" level=info msg="PodVM Created with mz2-2x16 Instance type successfully..."
time="2023-09-11T03:23:45-07:00" level=info msg="PodVM name: annotations-cpu-mem"
time="2023-09-11T03:23:45-07:00" level=debug msg="Instance number: 0, Instance id: 0797_cecb35d1-8c41-40fe-84b4-f291a53d5906, Instance name: podvm-annotations-cpu-mem-0336523b"
time="2023-09-11T03:23:45-07:00" level=info msg="Deleting pod annotations-cpu-mem..."
--- PASS: TestPodVMwithAnnotationsCPUMemory (25.63s)
    --- PASS: TestPodVMwithAnnotationsCPUMemory/PodVMwithAnnotationsCPUMemory_test (25.63s)
        --- PASS: TestPodVMwithAnnotationsCPUMemory/PodVMwithAnnotationsCPUMemory_test/PodVM_with_Annotations_CPU_Memory_is_created (0.43s)
=== RUN   TestPodVMwithAnnotationsInvalidInstanceType
=== RUN   TestPodVMwithAnnotationsInvalidInstanceType/PodVMwithAnnotationsInvalidInstanceType_test
=== RUN   TestPodVMwithAnnotationsInvalidInstanceType/PodVMwithAnnotationsInvalidInstanceType_test/Failed_to_Create_PodVM_with_Annotations_Invalid_InstanceType
time="2023-09-11T03:23:51-07:00" level=info msg="Got Expected Error: Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : failed to verify instance type: requested instance type (\"bz2-2x4\") is not part of supported instance types list: unknown"
time="2023-09-11T03:23:51-07:00" level=info msg="Deleting pod annotations-invalid-instance-type..."
--- PASS: TestPodVMwithAnnotationsInvalidInstanceType (5.33s)
    --- PASS: TestPodVMwithAnnotationsInvalidInstanceType/PodVMwithAnnotationsInvalidInstanceType_test (5.33s)
        --- PASS: TestPodVMwithAnnotationsInvalidInstanceType/PodVMwithAnnotationsInvalidInstanceType_test/Failed_to_Create_PodVM_with_Annotations_Invalid_InstanceType (0.19s)
=== RUN   TestPodVMwithAnnotationsLargerMemory
=== RUN   TestPodVMwithAnnotationsLargerMemory/PodVMwithAnnotationsLargerMemory_test
=== RUN   TestPodVMwithAnnotationsLargerMemory/PodVMwithAnnotationsLargerMemory_test/Failed_to_Create_PodVM_with_Annotations_Larger_Memory
time="2023-09-11T03:23:56-07:00" level=info msg="Got Expected Error: Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : failed to get instance type based on vCPU and memory annotations: no instance type found for the given vcpus (2) and memory (18432): unknown"
time="2023-09-11T03:23:56-07:00" level=info msg="Deleting pod annotations-too-big-mem..."
--- PASS: TestPodVMwithAnnotationsLargerMemory (5.36s)
    --- PASS: TestPodVMwithAnnotationsLargerMemory/PodVMwithAnnotationsLargerMemory_test (5.36s)
        --- PASS: TestPodVMwithAnnotationsLargerMemory/PodVMwithAnnotationsLargerMemory_test/Failed_to_Create_PodVM_with_Annotations_Larger_Memory (0.20s)
=== RUN   TestPodVMwithAnnotationsLargerCPU
=== RUN   TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test
=== RUN   TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test/Failed_to_Create_PodVM_with_Annotations_Larger_CPU
time="2023-09-11T03:24:01-07:00" level=info msg="Got Expected Error: Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : failed to get instance type based on vCPU and memory annotations: no instance type found for the given vcpus (3) and memory (12288): unknown"
time="2023-09-11T03:24:01-07:00" level=info msg="Deleting pod annotations-too-big-cpu..."
--- PASS: TestPodVMwithAnnotationsLargerCPU (5.39s)
    --- PASS: TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test (5.39s)
        --- PASS: TestPodVMwithAnnotationsLargerCPU/PodVMwithAnnotationsLargerCPU_test/Failed_to_Create_PodVM_with_Annotations_Larger_CPU (0.24s)
FAIL
FAIL	github.com/confidential-containers/cloud-api-adaptor/test/e2e	637.555s
FAIL
```

sudharshanibm3 commented:
Hi @stevenhorsman 
- The mount path for secret has been changed in recent CAA Logs
```
2023/09/11 08:25:06 [adaptor/proxy]         destination:/sys/fs/cgroup source:cgroup type:cgroup
2023/09/11 08:25:06 [adaptor/proxy]         destination:/etc/config source:/run/kata-containers/shared/containers/27265f3638f3f47f01451b217b41438939fbca7efdcaadf9df41260f5112c693-a094d1bc86f04bd8-config type:bind
2023/09/11 08:25:06 [adaptor/proxy]         destination:/sealed/etc/secret source:/run/kata-containers/shared/containers/27265f3638f3f47f01451b217b41438939fbca7efdcaadf9df41260f5112c693-4585b5f70e45b219-secret type:bind
```
- Here is the [Fix Code ](https://github.com/sudharshanibm3/cloud-api-adaptor/commit/8254e1ee32f24dc6676af44df7bb093808c7e6d2) for `TestCreatePodWithSecret `

stevenhorsman commented:
> Hi @stevenhorsman
> 
> * The mount path for secret has been changed in recent CAA Logs
> 
> ```
> 2023/09/11 08:25:06 [adaptor/proxy]         destination:/sys/fs/cgroup source:cgroup type:cgroup
> 2023/09/11 08:25:06 [adaptor/proxy]         destination:/etc/config source:/run/kata-containers/shared/containers/27265f3638f3f47f01451b217b41438939fbca7efdcaadf9df41260f5112c693-a094d1bc86f04bd8-config type:bind
> 2023/09/11 08:25:06 [adaptor/proxy]         destination:/sealed/etc/secret source:/run/kata-containers/shared/containers/27265f3638f3f47f01451b217b41438939fbca7efdcaadf9df41260f5112c693-4585b5f70e45b219-secret type:bind
> ```
> 
> * Here is the [Fix Code ](https://github.com/sudharshanibm3/cloud-api-adaptor/commit/8254e1ee32f24dc6676af44df7bb093808c7e6d2) for `TestCreatePodWithSecret `

Thanks, can you raise an issue for the secrets test failure, so we can track it into open-source please. Thanks!

stevenhorsman commented:
FYI Sudharshan, the libvirt tests are failing:
```
Error: test/e2e/libvirt_test.go:14:27: cannot use assert (variable of type LibvirtAssert) as CloudAssert value in argument to doTestCreateSimplePod: LibvirtAssert does not implement CloudAssert (missing method getInstanceType)
Error: test/e2e/libvirt_test.go:20:34: cannot use assert (variable of type LibvirtAssert) as CloudAssert value in argument to doTestCreatePodWithConfigMap: LibvirtAssert does not implement CloudAssert (missing method getInstanceType)
Error: test/e2e/libvirt_test.go:25:31: cannot use assert (variable of type LibvirtAssert) as CloudAssert value in argument to doTestCreatePodWithSecret: LibvirtAssert does not implement CloudAssert (missing method getInstanceType)
Error: test/e2e/libvirt_test.go:31:54: cannot use assert (variable of type LibvirtAssert) as CloudAssert value in argument to doTestCreatePeerPodContainerWithExternalIPAccess: LibvirtAssert does not implement CloudAssert (missing method getInstanceType)
Error: test/e2e/libvirt_test.go:37:32: cannot use assert (variable of type LibvirtAssert) as CloudAssert value in argument to doTestCreatePeerPodWithJob: LibvirtAssert does not implement CloudAssert (missing method getInstanceType)
Error: test/e2e/libvirt_test.go:42:41: cannot use assert (variable of type LibvirtAssert) as CloudAssert value in argument to doTestCreatePeerPodAndCheckUserLogs: LibvirtAssert does not implement CloudAssert (missing method getInstanceType)
Error: test/e2e/libvirt_test.go:47:44: cannot use assert (variable of type LibvirtAssert) as CloudAssert value in argument to doTestCreatePeerPodAndCheckWorkDirLogs: LibvirtAssert does not implement CloudAssert (missing method getInstanceType)
Error: test/e2e/libvirt_test.go:52:61: cannot use assert (variable of type LibvirtAssert) as CloudAssert value in argument to doTestCreatePeerPodAndCheckEnvVariableLogsWithImageOnly: LibvirtAssert does not implement CloudAssert (missing method getInstanceType)
Error: test/e2e/libvirt_test.go:57:66: cannot use assert (variable of type LibvirtAssert) as CloudAssert value in argument to doTestCreatePeerPodAndCheckEnvVariableLogsWithDeploymentOnly: LibvirtAssert does not implement CloudAssert (missing method getInstanceType)
Error: test/e2e/libvirt_test.go:62:70: cannot use assert (variable of type LibvirtAssert) as CloudAssert value in argument to doTestCreatePeerPodAndCheckEnvVariableLogsWithImageAndDeployment: LibvirtAssert does not implement CloudAssert (missing method getInstanceType)
Error: test/e2e/libvirt_test.go:62:70: too many errors
FAIL	github.com/confidential-containers/cloud-api-adaptor/test/e2e [build failed]
```
so I guess you need to add implementations/subs to the other CloudAssert implementations?

sudharshanibm3 commented:
> so I guess you need to add implementations/subs to the other CloudAssert implementations?

Yes, as we declared `getInstanceType` method in common `CloudAssert` interface so we can handle this problem with following two ways:
- If `libvirt` provisioner also uses flexible instance test cases, we can declare `getInstanceType` method inside `libvirt_test.go`
- else if this test cases are ibmcloud specific, then we can remove the `getInstanceType` method and make changes inside pre-existing `HasPodVM` method

stevenhorsman commented:
> > so I guess you need to add implementations/subs to the other CloudAssert implementations?
> 
> Yes, as we declared `getInstanceType` method in common `CloudAssert` interface so we can handle this problem with following two ways:
> 
> * If `libvirt` provisioner also uses flexible instance test cases, we can declare `getInstanceType` method inside `libvirt_test.go`
> * else if this test cases are ibmcloud specific, then we can remove the `getInstanceType` method and make changes inside pre-existing `HasPodVM` method

I would expect that other cloud-providers would like to have this functionality, but it won't be quite the same as this version, so potentially added stubs for `getInstanceType` to the other providers with a comment of their intent might be the best course of action and then we've check for them to review this?

sudharshanibm3 commented:
> Thanks, can you raise an issue for the secrets test failure, so we can track it into open-source please. Thanks!

Created an Issue in opensource for tracking sealed mount path for secret - #1424

sudharshanibm3 commented:
Hi @stevenhorsman 
Added stubs for `getInstanceType` to the other providers with a comment

stevenhorsman commented:
The libvirt tests are no just failing on the secret test which is a known-error, not related to this PR

wainersm commented:
> Given that common_suite_test.go is getting very large now, I wonder if we should consider splitting it up and whether, for example, these new flexible instance size tests should go into a separate tests file. @wainersm (or anyone else), do you have any thoughts on this?

I agree on separate the tests in thematic test files. An alternative would be to leave only the `doTest.*` methods in `common_suite_test.go` (making it a "one stop shop" of tests) but moving all auxiliary/implementation methods to somewhere else.

side question: has the e2e libvirt CI being useful?

stevenhorsman commented:
> side question: has the e2e libvirt CI being useful?

Yeah - it caught a problem with the CloudAssert not having an implementation for all providers

sudharshanibm3 commented:
Hi @stevenhorsman 
As per wainer's suggestion, can we move the [auxiliary functions](https://github.com/confidential-containers/cloud-api-adaptor/blob/b6376a2660d34a3d2a2290fe69ea5c1b1aa77168/test/e2e/common_suite_test.go#L427-L678) other than doTest* functions into a separate file called `helperFunc.go` something like that ?
Or we are only planning to isolate `instance type` test cases from `common_suite_test.go`?

sudharshanibm3 commented:
Hi @stevenhorsman ,

- Moved auxiliary functions and methods into helperFunctions.go and helperMethods.go files
- Created a separate commit for this change

stevenhorsman commented on test/e2e/helperMethods.go:
I think we might be able to come up with more descriptive names (which I'm not good at), for example test_case_builder.go ?

stevenhorsman commented on test/e2e/helperFunctions.go:
I think go has snake_case convention for file names

stevenhorsman commented on test/e2e/common_suite_test.go:
```diff
@@ -36,104 +31,6 @@ import (
-	return tc
-}
-
 func (tc *testCase) run() {
```
Should testCase run move out somewhere as well?

sudharshanibm3 replied:
Hi @stevenhorsman ,
I just moved run() method to different file, but there  [testEnv](https://github.com/confidential-containers/cloud-api-adaptor/pull/1420/commits/f2680539058854d92ba2954f3997ba2ce8a68a8b#diff-108219a173223ecaadfe6371b6295385a0fbe7278314ad4f7e768b2c3e9536a9R399) become `undecalred`, once I changed the file name with `_test` suffix, it got resolved.

stevenhorsman replied:
I think that's because go unit test files `_test.go` are compiled separately and will try to be executed as tests. I'm not a golang expert though, so someone else might be able to give better advice

stevenhorsman commented on test/e2e/builder_test.go:
Sorry, maybe we should discuss in the image, but this naming makes it sound like unit tests of the `builder.go` file, so is a bit misleading. I think even test_builder.go would be better

stevenhorsman commented:
BTW https://github.com/kubernetes-sigs/e2e-framework/blob/main/examples/suites/README.md might be helpful as an example if we decide to separate tests into separate files

sudharshanibm3 commented:
Hi @stevenhorsman 
- As per this [documentation](https://github.com/kubernetes-sigs/e2e-framework/blob/main/examples/suites/README.md), It seems in order to use the `testEnv` variable, the file name must be suffixed with `_test.go`
- Here are some filename suggestions for file that contains helper Functions - currently [helper_test.go](https://github.com/confidential-containers/cloud-api-adaptor/blob/f2680539058854d92ba2954f3997ba2ce8a68a8b/test/e2e/helper_test.go)
-- test_checkers.go
-- test_assertions.go
-- test_helpers.go
-- test_validators.go
-- common_helpers.go

- Here are some filename suggestion for file that contains helper Methods - currently [builder_test.go](https://github.com/confidential-containers/cloud-api-adaptor/blob/f2680539058854d92ba2954f3997ba2ce8a68a8b/test/e2e/builder_test.go)
(also testEnv works only with _test suffixed files)
-- assessment_method_test.go
-- helper_methods_test.go
-- builder_methods_test.go

stevenhorsman commented:
> Hi @stevenhorsman
> 
> * As per this [documentation](https://github.com/kubernetes-sigs/e2e-framework/blob/main/examples/suites/README.md), It seems in order to use the `testEnv` variable, the file name must be suffixed with `_test.go`
> * Here are some filename suggestions for file that contains helper Functions - currently [helper_test.go](https://github.com/confidential-containers/cloud-api-adaptor/blob/f2680539058854d92ba2954f3997ba2ce8a68a8b/test/e2e/helper_test.go)
>   -- test_checkers.go
>   -- test_assertions.go
>   -- test_helpers.go
>   -- test_validators.go
>   -- common_helpers.go
> * Here are some filename suggestion for file that contains helper Methods - currently [builder_test.go](https://github.com/confidential-containers/cloud-api-adaptor/blob/f2680539058854d92ba2954f3997ba2ce8a68a8b/test/e2e/builder_test.go)
>   (also testEnv works only with _test suffixed files)
>   -- assessment_method_test.go
>   -- helper_methods_test.go
>   -- builder_methods_test.go

Ok, if we believe that is needs the `_test` suffix, then that sounds okay. Thanks!

sudharshanibm3 commented:
Updated File names for auxiliary functions and methods
[helper_test.go](https://github.com/confidential-containers/cloud-api-adaptor/blob/f2680539058854d92ba2954f3997ba2ce8a68a8b/test/e2e/helper_test.go) --> assessment_helpers.go
[builder_test.go](https://github.com/confidential-containers/cloud-api-adaptor/blob/f2680539058854d92ba2954f3997ba2ce8a68a8b/test/e2e/builder_test.go) --> assessment_runner_test.go

stevenhorsman commented:
@sudharshanibm3 -  sorry I didn't get around to looking at this yet, it looks like you have a conflict and due to the nature of this PR moving a lot of test utils around I think it would be good to notify people on slack about it before it gets merged as rebasing active changes will be tricky

stevenhorsman reviewed (approved):
I re-tested this locally and it all passed and I think at this point it's probably better to get it in rather than iterate too much on the refactoring to try and get it perfect. Thanks for the code and patience!

wainersm reviewed (approved):
hi @sudharshanibm3 thank you so much on helping organize the code!
